### PR TITLE
New properties `Dataset.sizes` and `DataArray.sizes`

### PR DESCRIPTION
**Pull request by shoyer**
This allows for consistent access to dimension lengths on ``Dataset`` and
``DataArray``

xref #921 (doesn't resolve it 100%, but should help significantly)

### CODE REVIEW

File: xarray/core/common.py [hunk 1]
```diff
@@ -335,6 +335,36 @@ def pipe(self, func, *args, **kwargs):
...
+        squeezed : same type as caller
+            This object, but with with all or a subset of the dimensions of
+            length 1 removed.
+
+        See Also
+        --------
+        numpy.squeeze
+        """
+        if dim is None:
+            dim = [d for d, s in self.sizes.item() if s == 1]
```
Comment (MaximilianR) [inline xarray/core/common.py]: I think `items`?

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -110,6 +111,31 @@ def __setitem__(self, key, value):
         self.data_array[pos_indexers] = value
 
 
+class DataArraySizes(collections.Mapping, formatting.ReprMixin):
```
Comment (MaximilianR) [inline xarray/core/dataarray.py]: Does this need a separate class that holds the array? 
Could it be a `dict` of dim-name to dim-size?

Comment (shoyer) [inline xarray/core/dataarray.py]: Probably not -- could just generate an object on the fly, e.g., `Frozen(OrderedDict(zip(self.dims, self.shape)))`. Let me do a quick micro benchmark to see if that has appreciably different performance.

File: xarray/core/common.py (hunk 1 above)
Comment (shoyer) [inline xarray/core/common.py]: Oops indeed

**Comment by shoyer** [comment]
Updated to remove the custom class. It makes accessing a single dimension length from `DataArray.sizes` slightly slower, but within the range of usual Python overhead (2.5 vs. 5 us). Could add a cache at some later point if necessary.

**Pull request merged**